File: backend/shared/auth/jwt-utils.js
/**
 * JWT Utilities - Compartido entre servicios
 * 
 * ⚠️ SEGURIDAD:
 * - NO incluye credenciales hardcodeadas
 * - Usa AWS Systems Manager Parameter Store para secretos
 * - Compatible con AWS Academy (roles rotativos)
 */

const jwt = require('jsonwebtoken');
const { getParameter } = require('../utils/parameter-store');

let cachedSecret = null;

/**
 * Obtener JWT_SECRET desde Parameter Store (con cache)
 */
async function getJwtSecret() {
  if (cachedSecret) {
    return cachedSecret;
  }

  try {
    cachedSecret = await getParameter('/fridays/jwt-secret', true);
    return cachedSecret;
  } catch (error) {
    console.error('⚠️ No se pudo obtener JWT_SECRET de Parameter Store, usando fallback');
    // Fallback solo para desarrollo local
    return process.env.JWT_SECRET || 'dev-secret-key-fridays-2025';
  }
}

/**
 * Generar JWT token con tenant_id obligatorio
 */
async function generateToken(payload) {
  // Comparar case-insensitive para el rol Cliente
  const isCliente = payload.role && payload.role.toUpperCase() === 'CLIENTE';
  
  if (!payload.tenant_id && !isCliente) {
    throw new Error('tenant_id es obligatorio para usuarios staff');
  }

  const secret = await getJwtSecret();
  const expiresIn = process.env.JWT_EXPIRES_IN || '24h';

  return jwt.sign(payload, secret, { expiresIn });
}

/**
 * Verificar JWT token y validar tenant_id
 */
async function verifyToken(token) {
  try {
    const secret = await getJwtSecret();
    const decoded = jwt.verify(token, secret);

    // Validación obligatoria de tenant_id para staff (case-insensitive)
    const isCliente = decoded.role && decoded.role.toUpperCase() === 'CLIENTE';
    if (!isCliente && !decoded.tenant_id) {
      throw new Error('Token inválido: falta tenant_id');
    }

    return decoded;
  } catch (error) {
    console.error('❌ Token verification failed:', error.message);
    throw new Error('Invalid token');
  }
}

/**
 * Decodificar token sin verificar (útil para debugging)
 */
function decodeToken(token) {
  return jwt.decode(token);
}

/**
 * Extraer datos del usuario del event de API Gateway
 * (después de pasar por el authorizer)
 */
function getUserFromEvent(event) {
  const authorizer = event.requestContext?.authorizer;
  
  if (!authorizer) {
    return null;
  }

  return {
    userId: authorizer.userId,
    email: authorizer.email,
    role: authorizer.role,
    tenant_id: authorizer.tenant_id === 'null' ? null : authorizer.tenant_id
  };
}

/**
 * Validar que un usuario tenga permisos para acceder a un recurso
 */
function validateAccess(user, requiredRoles = []) {
  if (!user) {
    throw new Error('Usuario no autenticado');
  }

  // Validar tenant_id para staff
  if (user.role !== 'CLIENTE' && !user.tenant_id) {
    throw new Error('Acceso denegado: tenant_id requerido');
  }

  // Validar roles si se especifican
  if (requiredRoles.length > 0 && !requiredRoles.includes(user.role)) {
    throw new Error(`Acceso denegado: se requiere rol ${requiredRoles.join(' o ')}`);
  }

  return true;
}

module.exports = {
  generateToken,
  verifyToken,
  decodeToken,
  getUserFromEvent,
  validateAccess
};

File: backend/shared/auth/jwt_utils.py
"""
JWT Utilities
Maneja la creación y validación de tokens JWT usando secret de AWS Systems Manager Parameter Store
"""

import jwt
import os
import boto3
from datetime import datetime, timedelta
from functools import lru_cache

ssm = boto3.client('ssm')

# Cache del secret para evitar múltiples llamadas a Parameter Store
@lru_cache(maxsize=1)
def get_jwt_secret():
    """
    Obtiene el JWT secret de AWS Systems Manager Parameter Store
    """
    try:
        secret_name = os.environ.get('JWT_SECRET_PARAM', '/fridays/jwt-secret')
        response = ssm.get_parameter(Name=secret_name, WithDecryption=True)
        return response['Parameter']['Value']
    except Exception as e:
        print(f"Error obteniendo JWT secret: {str(e)}")
        raise Exception("No se pudo obtener el JWT secret")


def generate_token(user_id, email, role, tenant_id=None):
    """
    Genera un token JWT para un usuario
    """
    secret = get_jwt_secret()
    
    payload = {
        'userId': user_id,
        'email': email,
        'role': role,
        'tenantId': tenant_id,
        'iat': datetime.utcnow(),
        'exp': datetime.utcnow() + timedelta(days=7)  # Token válido por 7 días
    }
    
    token = jwt.encode(payload, secret, algorithm='HS256')
    return token


def decode_token(token):
    """
    Decodifica y valida un token JWT
    Retorna el payload si es válido, levanta excepción si no
    """
    secret = get_jwt_secret()
    
    try:
        payload = jwt.decode(token, secret, algorithms=['HS256'])
        return payload
    except jwt.ExpiredSignatureError:
        raise Exception("Token expirado")
    except jwt.InvalidTokenError:
        raise Exception("Token inválido")


def extract_token_from_header(authorization_header):
    """
    Extrae el token del header Authorization
    Formato esperado: "Bearer <token>"
    """
    if not authorization_header:
        raise Exception("No se proporcionó token de autorización")
    
    parts = authorization_header.split()
    
    if len(parts) != 2 or parts[0].lower() != 'bearer':
        raise Exception("Formato de token inválido. Use: Bearer <token>")
    
    return parts[1]
